packages/web/src/package/observers/invisible/InvisibleWrapper.ts

```diff
@@ -49,31 +63,42 @@ export function InvisibleWrapper({ fullKeyEncode }: Props): WrapperMiddleware {
     }
   }
 
-  function getMessage(message: string) {
-    if (message.length <= 4) {
-      const [valueCode] = stringToCodePoints(message);
-      return keyMemory.numberToValue(valueCode);
-    } else {
-      return message;
-    }
+  function retrieveMessages(text: string) {
+    return text
+      .split(MESSAGE_END)
+      .filter((m) => m.length)
+      .map((message) => {
+        if (message[0] === '{') {
+          // there is a json inside - the full key is included, not just number `fullKeyEncode`
+          return message;
+        } else {
+          const valueCode = Number(message);
+          return keyMemory.numberToValue(valueCode);
+        }
+      });
+  }
+
+  function encodeWithSeparator(message: string) {
+    return encodeMessage(message + MESSAGE_END);
```

Comment:
Maybe we can check here if the message already contains an END character and print a warning (only in some kind of debug mode). Or is this a bad idea?

Reply:
This is not facing the user directly, so I don't think it's necessary. The concern basically is that if someone implements a third party copy of this for e.g. PHP, he needs to know that newline is a separator, but we won't prevent that here, only by describing it well.